Performance and other improvements to serialization

dolio opened the pull request:
This PR contains improvements to the serialization of unison code/values. There are actually some improvements to the existing serialization, but some of the improvements require using a new format that is exposed with some new builtins.

- Performance has been improved. This includes the existing value (de)serializer. It just wasn't optimized as much as it could be.
- A new serialization format has been added. This is version 5 for values, and version 4 for code.
    - These are accessible via new `serialize.versioned` builtins. Deserialization automatically handles the new format if possible.
    - The implementations of this new format have been sectioned off in their own module, to try to get better optimization.
    - The major change to the format is that it contains a prefix of all `Reference` values used in the remainder, so that in the latter, just numbers can be used. This means that reference occurrences, which are ubiquitous if you're sending arbitrary data, shrink from 64 bytes to ~1 byte. So the format is likely much more compact.
    - `Map` has also been special cased, so that it doesn't get blown up to a full unison value, but efficiently serialized as a sequence of keys and values
    - All numbers in the new format are encoded using a variable width encoding to avoid wasting space
- As part of realizing the new format, some tricks with `StableName` have been implemented for canonicalizing values in memory. This is done somewhat opportunistically rather than uniformly. This enables using the same representation with `Reference` rather than refactoring, although the latter may be worth doing in the long run.
- Various unsafe FFI implementations have been replaced with safer `ForeignConvention` methods. It's still possible to get these wrong, but if a runtime representation has been changed, like switching from `Value` to `Referenced Value`, these FFI methods will automatically be updated, instead of having to track them down.

Existing functionality that relies on serialization, like hashing, has been left using the older formats. If they were changed to the new format, that would change the hash of _lots_ of stuff, so that'd be a significant breaking change. We could add a `versioned` analogue of those functions in the future, maybe. Also, the old `serialize` functions should work as before, so the new stuff is opt-in.

As for the performance improvements, there are basically 4 steps in serialization.

1. Going from a runtime value to an interchange value
2. Serializing an interchange value
3. Deserializing an interchange value
4. Going from an interchange value to a runtime value

I've been looking at a test case that involves serializing a list of 100,000 pairs of numbers.

1 has gotten slightly worse, because it's necessary to do a little more work for the improved format. However, the difference is like `50ms -> 65ms`, while the serialization steps are 3-5x that cost.

2 and 3 have been significantly improved, even for the old format. For my test case, 2 took ~500ms and 3 took ~700ms originally. Now 2 takes ~270ms for the old format and ~220ms for the new format, while 3 takes ~210ms for the old and ~150ms for the new.

4 used to take ~700ms (I think). There isn't much that can be done about this for the old format, because all references need to be looked up to numberings, and all the references are distinct in memory. Using the new format, this takes ~60ms instead, because of the sharing.

If you're sending `Map`s instead of lists of tuples, everything is even better. Like 30ms each for the serialization steps, and 75ms and 30ms for the deserialization steps.

The new formats are generally much smaller, too, because of all the sharing of references. The 64 byte hashes could easily be the bulk of the output, so you can easily get an order of magnitude better than the old format.

There are a couple scenarios that might incur some performance penalties with the current approach. Like if you call `Value.value` on something that contains a `Value` or `Code`, then some extra work might have to be done to ensure canonicity of the `Reference`s in memory. I've tried to avoid this work if possible, so that if everything happens to already be using the same reference, it will be extra fast. Also, I think this is probably not a common, super performance sensitive use case.

There are probably opportunities for marginal performance gains on top of the work here, but I'm not expecting large, general factors at this point. Specific gains for additional special cases might be possible.

Existing transcripts and unison tests pass. I haven't written new tests yet, but I'll get working on those once CI is rolling here.

dolio committed:
Use lazy bytestring for Value deserialization

The old version was using `toArray` to turn the unison bytes into a
form that can be consumed by the serializer. This seems to perform
quite badly. I think any concrete choice would be better, but I chose
lazy bytestring because it seems slightly better to produce a chunked
input from the rope that is already chunked.

dolio committed:
Don't copy the bytestring in `getHash`

Building the hash converts to a ShortByteString, which copies again,
so the copy just wastes time.

dolio committed:
Import get/putVarInt into Unison.Runtime.Serialize

This seems to allow some better inlining/specialization.

dolio committed:
Add some strategic inlines to boost performance

dolio committed:
Add a dedicated Seq deserializer

dolio committed:
Inline putFoldable

Seems to be a minor but detectable improvement

dolio committed:
Correct `putVarInt` and make `putLength` inlined

The former was missing a conversion to unsigned to avoid shifting
issues. This pre-step also factors it so that the core loop is inlined
and specialized, which is probably better.

Inlining `putLength` ensures that the `putVarInt` call is inlined and
specialized. This actually seems necessary.

dolio committed:
Add a type for tracking canonical references

dolio committed:
Use `Referenced Value` as runtime value representation

This also includes new serialization support for referenced values,
and support for canonicalizing values in memory associated to this.

dolio committed:
Specialize some functions in U.R.A.Serialize

dolio committed:
Evaluate before canonicalizing

Apparently `makeStableName` does not evaluate its argument, and the
name is sensitive to the exact pointer, not just the true 'value'
underlying it. So, evaluating is essential for getting reproducible
results.

dolio committed:
Separate term and type references in Referenced values

This allows us to avoid an expensive traversal to calculate all the
term and type references in a deserialized term. The traversal was
using a Set of references for deduplication, which is slow.

dolio committed:
Inline some canonicalizer functions

dolio committed:
Use runtime numbering for reflectValue

Previously it was using a canonicalizer to deduplicate References, but
it's a bit faster to use the existing numbering and keep track of
canonical values that way.

dolio committed:
Miscellaneous ValueV5 optimizations

dolio committed:
Share References in function(Un)Replacements

dolio committed:
Move numbered reference support out of ValueV5

dolio committed:
Export fromLazyByteString from Bytes module

dolio committed:
Add a special case for maps in the new serialization format

Intermediate values now just store the `toList` of a map. The old
serializer handles this by expanding in-situ. The new one just
serializes the list of pairs. This makes sending maps in the new
format considerably faster, while the overall costs are comparable for
the old format, just moved around.

dolio committed:
Add a builtin for accessing versioned serialization

dolio committed:
Tweak Cont representation and add Reference traversal functions

Cont was storing a Map for some reason, but there isn't much reason
for it. Switched to a list of pairs, with corresponding serializer
changes. The format is not any different, since maps were represented
as lists of pairs there.

Reference traversal functions traverse all the References in a Value,
Cont, etc. These are distinct from some of the traversals/folds that
were there before, because those traverse references needed for
loading.

dolio committed:
Add a note to the `Referenced` type

dolio committed:
Various serialization and representation changes

The runtime Code representation is now `Referenced`

This also includes a fleshed out V4 code format taking advantage of
this to store compacted references like the V5 value format.

Serialization has been updated to recognize the new version and call
into it.

Some canonicalization code has been tweaked to support more refined
canonicalization procedures. This is used when reflecting quoted
values and code, to canonicalize references across the whole value.

Various unsafe FFI marshalling has been replaced by safer
alternatives. They now make use of `BuiltinForeign` instances to make
sure that they are using a builtin reference with the right
corresponding type.

dolio committed:
Add a builtin for versioned code serialization

dolio committed:
Fix `Code` equality

dolio committed:
Fix cache_ decoding; use more 'safer' FFI conventions

dolio committed:
Transcript updates

dolio committed:
fix test compilation

dolio committed:
Get rid of separate CodeV3 and ValueV4 files

Stack pulls them in automatically, but I haven't been using them in my
cabal-based setup.

dolio committed:
automatically run ormolu

dolio commented on unison-runtime/src/Unison/Runtime/Serialize.hs:
```diff
@@ -110,7 +139,8 @@ getLength ::
   m n
-getLength = unVarInt <$> deserialize
+getLength = getVarInt
+{-# INLINE getLength #-}
```
BTW, this actually makes a difference.

I think what happens is that since `getVarInt` takes 0 args, and is marked inline, this gets expanded into a definition that is too big to inline on its own. So an inline pragma can make a difference on `x = y` definitions if `y` alone can inline to something big.

ceedubs commented:
Nice! I've confirmed that the Unison Cloud integration test suite passes with this branch of ucm. But that's without using the new format at all. I'll be on vacation for the next week and won't get a chance to give it a try until after that.

pchiusano reviewed (approved):
Nice!

I'd add round trip tests and an .ser file test.

dolio committed:
Fix `branchLinks` not traversing the `MatchNumeric` reference

I don't think this would ever cause problems before, because it would
always be a builtin reference. But it is necessary for
canonicalization.

dolio committed:
Canonicalize term/type link references during reflection

dolio committed:
Add v5 serialization to serialized tests

We can still generate v4 and v5 tests now.

Some v5 files were already checked in. I'm not sure what was in them.
At some point I thought we'd need a version bump, but then made things
work without it, so they are probably left over from that.

dolio committed:
automatically run ormolu
